src/librustc_lint/builtin.rs

```diff
@@ -1265,3 +1265,57 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UnionsWithDropFields {
         }
     }
 }
+
+
+/// Lint for casts from types other than `usize` or `isize` to raw pointers
+pub struct IntoToRawPtrCast;
+
+declare_lint! {
+    pub INT_TO_RAW_PTR_CAST,
+    Deny,
+    "cast from signed int other than `isize` to raw pointer"
+}
+
+impl LintPass for IntoToRawPtrCast {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(INT_TO_RAW_PTR_CAST)
+    }
+}
+
+impl<'a, 'tcx> LateLintPass<'a, 'tcx> for IntoToRawPtrCast {
+    fn check_expr(&mut self, cx: &LateContext, expr: &hir::Expr) {
+        if let hir::ExprCast(ref base, _) = expr.node {
+            let base_ty = cx.tables.expr_ty(base);
+            let target_ty = cx.tables.expr_ty(expr);
+            if let ty::TyRawPtr(..) = target_ty.sty {
+                match base_ty.sty {
+                    ty::TyInt(ast::IntTy::Is) => return,
+                    ty::TyInt(ast::IntTy::I128) |
+                    ty::TyInt(ast::IntTy::I64) => {
+                        if cx.tcx.data_layout.pointer_size.bytes() >= 8 {
```

Comment:
Why isn't this comparing the `ity` in `ty::TyInt(ity)` to `tcx.sess.target.int_type`?